File: lib/java/opentoken/src/main/java/com/truveta/opentoken/io/parquet/TokenParquetReader.java
/**
 * Copyright (c) Truveta. All rights reserved.
 */
package com.truveta.opentoken.io.parquet;

import java.io.IOException;
import java.util.HashMap;
import java.util.Map;
import java.util.NoSuchElementException;

import org.apache.hadoop.conf.Configuration;
import org.apache.hadoop.fs.Path;
import org.apache.parquet.example.data.Group;
import org.apache.parquet.hadoop.ParquetReader;
import org.apache.parquet.hadoop.example.GroupReadSupport;
import org.apache.parquet.schema.GroupType;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import com.truveta.opentoken.io.TokenReader;
import com.truveta.opentoken.processor.TokenConstants;

/**
 * Reads encrypted tokens from a Parquet file for decryption.
 * Expected columns: RuleId, Token, RecordId
 */
public class TokenParquetReader implements TokenReader {
    private static final Logger logger = LoggerFactory.getLogger(TokenParquetReader.class);
    
    private final ParquetReader<Group> reader;
    private Group currentGroup;
    private boolean hasNextCalled = false;
    private boolean hasNextResult = false;
    
    /**
     * Initialize the class with the input file in Parquet format.
     * 
     * @param filePath The input file path.
     * @throws IOException If an I/O error occurs.
     */
    public TokenParquetReader(String filePath) throws IOException {
        Configuration conf = new Configuration();
        Path path = new Path(filePath);
        
        GroupReadSupport readSupport = new GroupReadSupport();
        this.reader = ParquetReader.builder(readSupport, path).withConf(conf).build();
        
        // Read the first group to validate schema
        currentGroup = reader.read();
        if (currentGroup != null) {
            GroupType schema = currentGroup.getType();
            
            // Validate required columns
            boolean hasRuleId = false;
            boolean hasToken = false;
            boolean hasRecordId = false;
            
            for (org.apache.parquet.schema.Type field : schema.getFields()) {
                String fieldName = field.getName();
                if (TokenConstants.RULE_ID.equals(fieldName)) {
                    hasRuleId = true;
                } else if (TokenConstants.TOKEN.equals(fieldName)) {
                    hasToken = true;
                } else if (TokenConstants.RECORD_ID.equals(fieldName)) {
                    hasRecordId = true;
                }
            }
            
            if (!hasRuleId || !hasToken || !hasRecordId) {
                throw new IOException("Missing required columns: RuleId, Token, or RecordId");
            }
            
            hasNextCalled = true;
            hasNextResult = true;
        }
    }
    
    @Override
    public boolean hasNext() {
        if (hasNextCalled) {
            return hasNextResult;
        }
        
        try {
            currentGroup = reader.read();
            hasNextResult = currentGroup != null;
            hasNextCalled = true;
            return hasNextResult;
        } catch (IOException e) {
            logger.error("Error reading next group", e);
            hasNextResult = false;
            hasNextCalled = true;
            return false;
        }
    }
    
    @Override
    public Map<String, String> next() {
        if (!hasNextCalled || !hasNextResult) {
            if (!hasNext()) {
                throw new NoSuchElementException("No more rows available");
            }
        }
        
        hasNextCalled = false;
        
        Map<String, String> row = new HashMap<>();
        GroupType schema = currentGroup.getType();
        
        for (org.apache.parquet.schema.Type field : schema.getFields()) {
            String fieldName = field.getName();
            try {
                String value = currentGroup.getString(fieldName, 0);
                row.put(fieldName, value != null ? value : "");
            } catch (Exception e) {
                row.put(fieldName, "");
            }
        }
        
        return row;
    }
    
    @Override
    public void close() throws IOException {
        if (reader != null) {
            reader.close();
        }
    }
}

File: lib/python/opentoken/src/main/opentoken/io/parquet/token_parquet_reader.py
"""
Copyright (c) Truveta. All rights reserved.
"""

import logging
from typing import Iterator, Dict

try:
    import pyarrow.parquet as pq
except ImportError:
    raise ImportError("pyarrow is required for Parquet support. Install with: pip install pyarrow")

from opentoken.io.token_reader import TokenReader
from opentoken.processor.token_constants import TokenConstants


logger = logging.getLogger(__name__)


class TokenParquetReader(TokenReader):
    """
    Reads encrypted tokens from a Parquet file for decryption.
    Expected columns: RuleId, Token, RecordId
    """

    def __init__(self, file_path: str):
        """
        Initialize the class with the input file in Parquet format.

        Args:
            file_path: The input file path.

        Raises:
            IOError: If an I/O error occurs.
        """
        try:
            self.file_path = file_path
            self.parquet_file = pq.ParquetFile(file_path)
            self.table = self.parquet_file.read()
            self.current_row = 0
            self.total_rows = len(self.table)
            
            # Validate required columns
            required_columns = {TokenConstants.RULE_ID, TokenConstants.TOKEN, TokenConstants.RECORD_ID}
            column_names = set(self.table.column_names)
            if not required_columns.issubset(column_names):
                missing = required_columns - column_names
                raise ValueError(f"Missing required columns: {missing}")

        except Exception as e:
            logger.error(f"Error in reading Parquet file: {e}")
            raise IOError(f"Failed to read Parquet file: {file_path}") from e

    def __iter__(self) -> Iterator[Dict[str, str]]:
        """
        Iterate over rows in the Parquet file.

        Returns:
            Iterator of dictionaries with RuleId, Token, and RecordId.
        """
        return self

    def __next__(self) -> Dict[str, str]:
        """
        Get the next row from the Parquet file.

        Returns:
            Dictionary with RuleId, Token, and RecordId.
        """
        if self.current_row >= self.total_rows:
            raise StopIteration

        row_dict = {}
        for column_name in self.table.column_names:
            value = self.table.column(column_name)[self.current_row].as_py()
            row_dict[column_name] = str(value) if value is not None else ''

        self.current_row += 1
        return row_dict

    def close(self):
        """Close the file handle."""
        # PyArrow ParquetFile doesn't need explicit closing
        pass

    def __enter__(self):
        """Context manager entry."""
        return self

    def __exit__(self, exc_type, exc_val, exc_tb):
        """Context manager exit."""
        self.close()
